EZP-26007: Use QueryController for the Home page and Blog

clash82 opened the pull request:
Link to Jira task: https://jira.ez.no/browse/EZP-26007

**This PR should be merged only on top of ezpublish-kernel v6.4.2**

Discovered during the implementation:
- https://jira.ez.no/browse/EZP-26086 (fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1735)

clash82 commented:
@bdunogier there is an issue with QueryType controller

have a look at this line https://github.com/ezsystems/ezplatform-demo/pull/23/files#diff-f5fc7a0c95a8ac0821192d8eb102a87eR24 - I cannot use parameter here because it throws me

```
Attempted to load class "%app.query_type.latest_blog_posts.class%" from the global namespace.
Did you forget a "use" statement?
```

when I use class name `AppBundle\QueryType\LatestBlogPostsQueryType` then it works fine

bdunogier commented:
Hmmm, it doesn't use the QueryController, but Query Types, does it not ?

bdunogier commented:
About the class name, I thought I have handled this already, but maybe not. The QueryTypePass does stuff on the QueryType class, but it was supposed to resolve parameters.

But you could just name your QueryType by convention, and let the pass do the rest (if a QueryType class is in `<bundlename>/QueryType/NameQueryType.php`, it gets tagged automatically.

clash82 commented:
> Hmmm, it doesn't use the QueryController, but Query Types, does it not ?

That's true, it's only the first step :-)

> But you could just name your QueryType by convention, and let the pass do the rest (if a QueryType class is in <bundlename>/QueryType/NameQueryType.php, it gets tagged automatically.

I can move it to dedicated directory but I wanted to give you a feedback about the exception I have.

andrerom commented:
@clash82 Any suggestions on how we can avoid the given exception for newcomers?

clash82 commented:
> Any suggestions on how we can avoid the given exception for newcomers?

I'm afraid not. This should be fixed in `QueryTypePass.php` file (https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Bundle/EzPublishCoreBundle/DependencyInjection/Compiler/QueryTypePass.php#L31).

Possible solution might be:
```php
...

$queryTypeClass = $queryTypeDefinition->getClass();

if (strpos($queryTypeClass, '%') !== false) {
    $queryTypeClass = $container->getParameterBag()->resolveValue($queryTypeClass);
}
```

...but that's not all. When I'm trying to inject languages into `QueryType` service using callers (more details here: https://github.com/ezsystems/ezplatform-demo/pull/23/files#diff-f5fc7a0c95a8ac0821192d8eb102a87eR24) then instead of languages I'm getting null.

This works fine for any regular services but not with QueryTypes as a services.

clash82 commented:
Ready for final review @andrerom @bdunogier (after getting back from holidays...) @damianz5 @SylvainGuittard 

Sad thing about the `QueryType as a service` is that (at least for now) it's impossible to inject anything into this service using setters or constructor 😢

bdunogier commented:
> Sad thing about the QueryType as a service is that (at least for now) it's impossible to inject anything into this service using setters or constructor 😢

Uh ? I have done tha

clash82 commented:
> Uh ? I have done it myself, it is supposed to work... Did you report an issue about it, with details of what you've tried ?

Here you are https://jira.ez.no/browse/EZP-26086

Should I also report issue with parameters (mentioned few lines above...) or this will be fixed in one go?

clash82 commented:
Ready for final review @bdunogier @andrerom @damianz5 @kamilmusial

kamilmusial commented:
+1

damianz5 commented:
+1

andrerom commented:
+1

clash82 commented:
@bdunogier I'd love to receive feedback from your side. If the approach here is good we will continue with rest of the controllers.

bdunogier commented on app/Resources/views/full/home_page.html.twig:
```diff
@@ -46,14 +42,10 @@
-                            })) }}
-                        {% endfor %}
-                    {% endif %}
+                    {{ render_esi(controller('ez_content:viewLocation', {
```
Reminder: viewLocation is deprecated. You should use viewContent, with both `contentId` and `locationId` as arguments.

bdunogier commented on app/Resources/views/full/blog_post.html.twig:
```diff
@@ -51,36 +51,10 @@
-                                    </div>
-                                </div>
-                            {% endfor %}
+                            {{ render_esi(controller('ez_content:viewLocation', {
```
Same as below about location view being deprecated.

bdunogier commented on src/AppBundle/eZ/Publish/Core/MVC/Symfony/Controller/Content/QueryController.php:
```diff
@@ -0,0 +1,64 @@
+ * A content view controller that runs queries based on the matched view configuration
+ * and support pagination using Pagerfanta as a pagination provider.
+ */
+class QueryController
```
I'd give it a more explicit name, like `PagerFantaQueryController`.

bdunogier commented on src/AppBundle/Resources/config/services.yml:
```diff
@@ -1,10 +1,28 @@
+        tags:
+            - { name: twig.extension }
+
+    app.query_controller:
```
Same than the controller's name. What about `app.paginated_query_controller` ?

bdunogier commented:
Looks good. Maybe too many changes in one commit, but that's all.

I'm just a little confused by the paginated controller: did you intend to have it override the one from the Core (hence its namespace) or not (in which case you could move it to a simpler, clearer path).

clash82 commented:
Updated, thanks @bdunogier 

> I'm just a little confused by the paginated controller: did you intend to have it override the one from the Core (hence its namespace) or not (in which case you could move it to a simpler, clearer NS).

I was not sure where should I put this controller so I decided to follow kernel path which (I thought) will be more understandable for developers. After your comment I moved it to the new location with other controllers, it looks less complicated now.

bdunogier commented on app/Resources/views/full/blog.html.twig:
```diff
@@ -37,15 +38,15 @@
                 <div class="{% if loop.index % 2 == 1 %}col-md-offset-2 {% endif %}col-md-5 space-top">
-                    {{ render_esi(controller('ez_content:viewLocation', {
+                    {{ render_esi(controller('ez_content:viewContent', {
                         locationId: post.contentInfo.mainLocationId,
```
Even though the `contentId` is an optional argument, I'd still specify it, along with the locationId:

```
{
  locationId: post.contentInfo.mainLocationId,
  contentId: post.contentInfo.contentId,
  viewType: 'line'
}
```

clash82 replied:
Should I always use both `locationId` and `contentId` when it's possible? documentation doesn't say nothing about this approach.

bdunogier replied:
Interesting question.

The issue with only using `contentId` is that any content item might have several locations. If you don't specify the `locationId`, the main one will be used, and it is probably not what you want.

The only problem with `locationId` alone is that it will be used to determine the `contentId`, which might be an issue in case of locationless content. But in the context of a search like you're using now, it shouldn't be a problem.

But on top of that, after checking the code again, _both_ `locationView` and `contentView` are deprecated in favour of `viewAction`. Can you replace those calls with `viewAction` ?

clash82 replied:
ok, I've got it now, thank you.
`viewAction` applied where needed.

bdunogier replied:
Great. Does it work as expected ?

clash82 replied:
Yes, works very good.
